sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java

```diff
@@ -0,0 +1,190 @@
+            "SentryAndroidOptions is required");
+
+    this.hub = Objects.requireNonNull(hub, "Hub is required");
+
+    this.options
+        .getLogger()
+        .log(
+            SentryLevel.DEBUG,
+            "ActivityLifecycleIntegration enabled: %s",
+            this.options.isEnableActivityLifecycleBreadcrumbs());
+
+    performanceEnabled = isPerformanceEnabled(this.options);
+
+    if (this.options.isEnableActivityLifecycleBreadcrumbs() || performanceEnabled) {
+      application.registerActivityLifecycleCallbacks(this);
+      options.getLogger().log(SentryLevel.DEBUG, "ActivityLifecycleIntegration installed.");
+    }
+  }
+
+  private boolean isPerformanceEnabled(final @NotNull SentryAndroidOptions options) {
+    return ((options.getTracesSampleRate() != null || options.getTracesSampler() != null)
+        && options.isEnableAutoActivityLifecycleTracing());
+  }
+
+  @Override
+  public void close() throws IOException {
+    application.unregisterActivityLifecycleCallbacks(this);
+
+    options.getLogger().log(SentryLevel.DEBUG, "ActivityLifecycleIntegration removed.");
+  }
+
+  private void addBreadcrumb(final @NonNull Activity activity, final @NotNull String state) {
+    if (options.isEnableActivityLifecycleBreadcrumbs()) {
+      final Breadcrumb breadcrumb = new Breadcrumb();
+      breadcrumb.setType("navigation");
+      breadcrumb.setData("state", state);
+      breadcrumb.setData("screen", getActivityName(activity));
+      breadcrumb.setCategory("ui.lifecycle");
+      breadcrumb.setLevel(SentryLevel.INFO);
+      hub.addBreadcrumb(breadcrumb);
+    }
+  }
+
+  private @NotNull String getActivityName(final @NonNull Activity activity) {
+    return activity.getClass().getSimpleName();
+  }
+
+  private void startTracing(final @NonNull Activity activity) {
+    if (performanceEnabled && !isRunningTransaction(activity)) {
+      final ITransaction transaction =
+          hub.startTransaction(getActivityName(activity), "navigation");
+
+      // lets bind to the scope so other integrations can pick it up
+      hub.configureScope(
+          scope -> {
+            // we we'd not like to overwrite existent transactions bound to the Scope manually.
+            if (scope.getTransaction() == null) {
+              scope.setTransaction(transaction);
+            }
+          });
+
+      activitiesWithOngoingTransactions.put(activity, transaction);
+    }
+  }
+
+  private boolean isRunningTransaction(final @NonNull Activity activity) {
+    return activitiesWithOngoingTransactions.containsKey(activity);
+  }
+
+  private void stopTracing(
+      final @NonNull Activity activity, final boolean enableAutoActivityLifecycleTracingFinish) {
+    if (performanceEnabled && enableAutoActivityLifecycleTracingFinish) {
+      final ITransaction transaction = activitiesWithOngoingTransactions.get(activity);
+      if (transaction != null && !transaction.isFinished()) {
```

Comment:
ideally `finish` does it atomically, so we don't fish twice, use AtomicBoolean and flip it